assembly/src/assembler/mod.rs

```diff
@@ -595,86 +601,148 @@ impl Assembler {
+                            .ensure_node(first_repeat_node)?;
+
+                        body_node_ids.push(first_repeat_node_id);
+                        for _ in 0..(*count - 1) {
+                            body_node_ids.push(repeat_node_id);
+                        }
+                    } else {
+                        for _ in 0..*count {
+                            body_node_ids.push(repeat_node_id);
+                        }
                     }
                 },
 
                 Op::While { body, .. } => {
-                    if let Some(basic_block_id) =
-                        basic_block_builder.make_basic_block(mast_forest_builder)?
-                    {
-                        node_ids.push(basic_block_id);
+                    if let Some(basic_block_id) = block_builder.make_basic_block()? {
+                        body_node_ids.push(basic_block_id);
                     }
 
-                    let loop_body_node_id =
-                        self.compile_body(body.iter(), proc_ctx, None, mast_forest_builder)?;
+                    let loop_node_id = {
+                        let loop_body_node_id = self.compile_body(
+                            body.iter(),
+                            proc_ctx,
+                            None,
+                            block_builder.mast_forest_builder_mut(),
+                        )?;
+                        block_builder.mast_forest_builder_mut().ensure_loop(loop_body_node_id)?
+                    };
+                    if let Some(decorator_ids) = block_builder.drain_decorators() {
+                        block_builder
+                            .mast_forest_builder_mut()
+                            .set_before_enter(loop_node_id, decorator_ids)
+                    }
 
-                    let loop_node_id = mast_forest_builder.ensure_loop(loop_body_node_id)?;
-                    node_ids.push(loop_node_id);
+                    body_node_ids.push(loop_node_id);
                 },
             }
         }
 
-        if let Some(basic_block_id) =
-            basic_block_builder.try_into_basic_block(mast_forest_builder)?
-        {
-            node_ids.push(basic_block_id);
-        }
+        let maybe_post_decorators: Option<Vec<DecoratorId>> =
+            match block_builder.try_into_basic_block()? {
+                BasicBlockOrDecorators::BasicBlock(basic_block_id) => {
+                    body_node_ids.push(basic_block_id);
+                    None
+                },
+                BasicBlockOrDecorators::Decorators(decorator_ids) => {
+                    // the procedure body ends with a list of decorators
+                    Some(decorator_ids)
+                },
+                BasicBlockOrDecorators::Nothing => None,
+            };
+
+        let procedure_body_id = if body_node_ids.is_empty() {
+            // We cannot allow only decorators in a procedure body, since decorators don't change
```

Comment:
This should probably be documented in the Miden documentation, specifically which instructions are decorators, and the fact that they cannot appear in a procedure by themselves.

I'm not sure how likely it is in practice, but I imagine the error would be pretty confusing.

Reply:
This can probably go into the [Procedures](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/code_organization.html#procedures) section. And we can refer to the a [advice injectors](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/io_operations.html#nondeterministic-inputs) as well as list `emit`, `debug`, and `trace` instructions (so that we don't have to list all advice injectors).

Reply:
Updated the docs [2e8481e](https://github.com/0xPolygonMiden/miden-vm/pull/1466/commits/2e8481e1b138c21eb31be2fc43021c9277866dc8). With #1457 we can fix it more holistically - the "decorators" name could be removed in favor of "debug instructions" or whatnot, and here in the "debug instructions" section we can state that these do not modify the MAST root, can be stripped away without affecting the behavior of the program, and cannot be the only instructions in a procedure